### ISSUE TITLE
Ensure we have the minimum setup for evaluating Display Ad success

### ISSUE THREAD

**Issue by mirie** [forem/forem#18380]
## Problem Statement

It is possible that we have a gap with our event tracking on impressions/engagements for anonymous users and/or that our success tracking system for logged-in users is lacking critical information. 

This is an issue to note this possible gap and to allow for some investigation. We may need to do add something to aid us in understanding the effectiveness of our Display Ads for both anonymous and logged in users.

Linking a related work regarding [product instrumentation](https://github.com/forem/forem-internal-eng/issues/511).

## Definition of done
- [ ] We understand our gaps in terms of being able to effectively evaluate the success of our efforts in regards to activation and retention
- [ ] We at minimum address the gaps so that we can better evaluate our efforts for [insert issue]
- [ ] If there are things to consider for the future as we continue building the Display Ad functionality out, we capture that in separate issues for future prioritization

### Supporting documentation
- Looking at `/admin/customization/display_ads` we have a `Success Rate`

#### Questions/Point of contact
- @mirie created this issue just so we could ask ourselves this question. It may not be an issue. You can ask @mirie for clarification on the ask detailed here
- @pkfrank is the Product owner for this overall work for the collection of issues under the linked Milestone.

**Comment by jaw6** [comment]
1. We *are* tracking post-article display ads via the same mechanism as we have had for sidebar display ads (that is: if a user is logged-in, we record an `impression`; if the user is logged-in and clicks on any link in the ad, we record a `click`)
2. We are *not* tracking impressions _or_ clicks for logged-out users
3. `success = clicks / impressions`
4. These are recorded in our app db, not in an external service (contrast with: [sponsored post tracking](https://github.com/forem/forem/commit/5baa086bd0b5f6441d8797385dc88831aa8a806b), which uses Google Analytics and has been updated to GA4)
5. This only-track-logged-in-users has been present since nearly the beginning of display ads (https://github.com/forem/forem/pull/3589)
6. I'm not sure if this is the _only_ reason, but _one_ reason is that our `DisplayAdEvent` table requires a `user_id` (which would not exist for a non-logged-in user)

:link: This will impact #18390 `cc` @Ridhwana

**Comment by Ridhwana** [comment]
I think this does a really good job of summarizing what we currently have in place and where our gaps lie in tracking Display Ads. ☝️


> We may need to do add something to aid us in understanding the effectiveness of our Display Ads for both anonymous and logged in users.

As per Joshuas findings,  we are not tracking impressions or clicks for logged-out users, hence when building out [the separation of the logged in and logged out user groups for Display Ads](https://github.com/forem/forem/issues/18390) I'd not focussed on the analytics.


Based on the definition of done for this ticket: 

>  We understand our gaps in terms of being able to effectively evaluate the success of our efforts in regards to activation and retention
>  We at minimum address the gaps so that we can better evaluate our efforts for [insert issue]
>  If there are things to consider for the future as we continue building the Display Ad functionality out, we capture that in separate issues for future prioritization 

Should we be addressing analytics for logged out users as part of this cycle?

**Comment by pkfrank** [comment]
Excited that we're summarizing the current state of things and identifying these important considerations.

> Should we be addressing analytics for logged out users as part of this cycle?

I'd love for us to shoot for getting this implemented during this cycle, if that seems doable.

**Comment by benhalpern** [comment]
I think the pattern here needs to follow just like the way we do logged in/non logged pageviews.

That is:

- We track _all_ signed in page views, and provide them a `user_id`
- For non-signed-in pageviews, we track one in every 10 events, which is less precision, but also less "waste".
- We assign all page views a `counts_for_number_of_views` number. Signed in page views "count" for 1, because we're tracking all of them. Signed out page views "count" for 10, because we're tracking 1 in 10.

I believe following this exact pattern is the way to go here, and should allow somewhat minimal big-picture changes. It would, of course, require some adjustments from count to sum in some cases and changes to how the ultimate success rate is factored, but that should all be pretty intuitive and testable.

Does this all make sense? Anything I'm missing?

**Comment by jaw6** [comment]
We will need a migration to remove foreign-key constraints and make `user_id` optional on `display_ad_events`. The history of this is potentially interesting:

* There _was_ a `counts_for` column on `display_ad_event`: https://github.com/forem/forem/pull/6806
* `user_id` _was_ originally created as optional, but we later added foreign_key constraints: https://github.com/forem/forem/pull/9786
* The motivation seems to have been cleaning up orphaned rows: https://github.com/forem/forem/pull/9765

My read here is that, circa 2020, we were concerned about the performance effects of this table, and that (also?) we deliberately moved _away_ from using `counts_for`, but it isn't clear why we made _that_ decision at the time. (Possibly: we intended to use `counts_for`, but didn't (?), and simply removed what wasn't actively used.)

Anyway, at the time of that last PR, there were ~300k orphan rows (including, presumably, both "display_ad orphans" and "user orphans").

That exact blazer query no longer works, but I'd be curious to run a count of `display_ad_events`. Maybe we want to estimate what an acceptable number of rows (/columns) might look like for this table? (Since our history suggests this is something we should be working to limit, and removing these constraints will push it in the bigger direction.) Or, is that not an active concern at this point?

**Comment by jaw6** [comment]
So, I'm just making some thoughts/notes here about the performance effects of these changes. From my perspective, there's maybe two concerns:

1. performance of a query to remove the `user_id` foreign-key constraint from `display_ad_events` 
2. performance over time of the `display_ad_events` table, as we generate more rows in that table

For 1: `display_ad_events` is currently 9,453,679 rows.

For 2: Adding display ads to post pages and tracking logged-out users seems like we'll see this table's grow at a rate similar to `page_views` (which already uses the `counts_for` method Ben outlines). `page_views` is currently 110,307,576. In the last ~month, we added ~3m rows to  `page_views`  and ~475k rows to `display_ad_events`.

With current traffic trends, the combination of DisplayAds on articles pages and logging events for non-logged-in users might increase the growth of this table ~3m rows, 10x its current growth rate, or a +30% row increase in the first month after introducing the changes.

I know "row count" is a pretty shallow tool (number of columns, complexity of relations, etc.) but at a first pass, it's a place to start. In any case, for this comparison, `page_views` and `display_ad_events` ought to be reasonably similar, design-wise.

Are there concerns if the performance of the `display_ad_events` table begins to resemble closely match the performance of the `page_views` table?

I will also investigate whether we have queries that leverage the `user_id` index -- the only one I know of currently is the `dependent` cascade when a user record is deleted.

**Comment by mirie** [comment]
Okay i’ve caught up on both Github and [this thread](https://forem-team.slack.com/archives/C7GE84DEJ/p1662112336305909) in Slack.

It looks like in order to understand conversion of display ads for anonymous users, we need to make a code change. There are some considerations to delve into regarding storage of event data -- this is related to how we do tracking just generally speaking.

Questions:
(1) If we follow the pattern of page view tracking, what do we have to keep in mind/allocate time for  (keeping/changing) re: overall event tracking implementation?
(2) If we are unsure of the answer for ^, what could we put in place that will get us some imperfect signaling so that we can at least get the logged out / logged in ads out to play around with, knowing that we have a fast follow for getting more accurate conversion data
(3) Finally, considering the tech debt raised in Slack, what's a reasonable way of paying this down?

**Comment by maestromac** [comment]
We currently use up 200GB of 2TB of our Heroku PostgreSQL plan. Storage wise it will be okay. That being said, page_views is taking up way too many roles and we are due for a clean-up there. I am worried the same will happy to us tracking display ads.

Having spoken to @jaw6 , because of how much feature demand and the possibility of data bloat, I think we ought to give using Ahoy fully for tracking the display Ads events and replace our own display Ads tracking. There are a few reasons for this.
1. Ahoy provides extra metadata such as geolocation and browser information.
2. Ahoy provides "rollup" feature which will allow us to properly deal with old data.
3. Ahoy is already part of the app, and this implementation should be pretty straightforward. We can do more with less code.

Overall, at the very least we should have a proof of concept. Implement Ahoy tracking on DisplayAd then use a week or a month of data to compare it with our own internal impression tracking, then maybe we'll make a better decision. I also think there's no better time than now to figure this out, as it can be much harder down the line.

**Comment by benhalpern** [comment]
@maestromac I think that is a good idea, but let me try and synthesize a couple concepts from this conversation and [some of what is going on in the other convo about data conventions](https://github.com/forem/forem-internal-eng/issues/511). I think my ideas are in line with what has been talked about, but let me know if I'd be off base. cc: @jaw6 @Ridhwana

Here is what I'm thinking:

- So in the conversation, @pkfrank brought up the idea that Ahoy events be generally restricted to data which _can_ be _ephemeral_, i.e. data structured such that if we need to pivot, we can do so flexibly in order to gather readable data to help solve problems for product development. That means Ahoy probably _should not_ be used for any data gathering which is not for analytics purposes, but is actually for the app to then take and use as further application logic. (i.e. we can't remove old pageviews because they need to be apples-to-apples forever.
- The topic of "rollup" is absolutely poignant here. And I think that is exactly what we need to be doing on our _native business logic tables_, like `page_views`, etc. I think we need to be _rolling up_ old pageviews where possible to shift from "counts for 10" to "counts for 100" on older data when there is enough total data to roll up. I believe we can take this similar pattern, but safely apply it to our more custom tables, _and keep Ahoy mostly for ephemeral analytics_.
- That would mean that we go through with the _adding `counts_for` to `display_ad_events`_ as described — and while probably doesn't have to be a blocker — we deal with bloat by envoking the "roll up" _pattern_, but re-implementing it safely in a way which tightly matches the business logic we need (I imagine this will be re-usable across pageviews and display ad events).

@jaw6 to get to your original questions, I think the reason the behavior was once designed to be going in this direction, and then changed away from that, is a matter of non-cohesion as far as following through on the originally intended behavior. More a matter of communication dynamics than a strictly close-knit tactic.

There is room for this to be different depending on a possible shift in where we land on the overall analytics discussion, but I personally believe there is merit to using Ahoy strictly for tracking and analytics which are not further ingested by application logic, and sticking to custom tables for tracking core data which needs to be continuously referenced within the flow of core app logic.

***

Final thought, @jaw6 : Correct me if I'm wrong, but I don't see an actual _db-level_ constraint on the `user_id` being present — just an app logic validation. Seems like if we add `optional: true`, we'd be good on that constraint. Again, let me know if I'm missing something, but I'm pretty sure the DB itself is totally cool with null `user_id` on this table.

**Comment by jaw6** [comment]
@benhalpern I'm not 100% sure I understand the distinction you're drawing here. I think you are saying that `page_views` are ingested by the application logic and should therefore be first-class domain objects, and I see where `page_views` are used for things like feed experiments, but I'm not seeing where `display_ad_events` are being ingested by the application?

---

> Correct me if I'm wrong, but I don't see an actual db-level constraint on the user_id being present — just an app logic validation. Seems like if we add optional: true, we'd be good on that constraint. Again, let me know if I'm missing something, but I'm pretty sure the DB itself is totally cool with null user_id on this table.

@benhalpern You are right on this, and I was wrong. The `foreign_key` on `display_ad_events.user_id` does **not** require `user_id` to be non-null. This is a big relief, actually; I had been worried about migrating that large table to remove that constraint.

**Comment by benhalpern** [comment]
@jaw6 — `display_ad_events` data is rolled up into `success_rate`, `clicks_count`, and `impressions_count` on the `display_ads` model. And then `success_rate` is rolled up into the algorithm which is used to generate which ads to show more often — i.e. more "successful" ads are shown more often, where success is measured here by clickthrough rate.

Generally as we're looking to make this functionality a core building block for platform mechanics, these specialized models have some merit — as opposed to if this were a fringe functionality which was not already in place and integrated.

**Comment by benhalpern** [comment]
@jaw6 PR looks great! Let me know if I can be helpful with the described next step once this lands.

**Comment by mirie** [comment]
I'm not sure why the PR isn't linked but adding it here: https://github.com/forem/forem/pull/18460 -- looks like we just need one more review / feedback before closing this one out.

**Comment by mirie** [comment]
Moving this to Blocked / In Progress as we want to wait until @maestromac returns so we can get a final look and be more in a ready state in case we notice an issue after merging. There's no immediate fire / need to move on this before Mac returns from vacation. :)

**Comment by mirie** [comment]
@maestromac / @jaw6: Should we plan to carry this over or are y'all chatting tomorrow (Friday)?

**Comment by jaw6** [comment]
:ok: #18508 has merged. I'm following along to make sure things are running, but we won't really see a lot of activity until about a week from now. I think we can close this issue as resolved and - if we should encounter any problems later - we can open a new issue.
